src/enact/types.py

```diff
@@ -0,0 +1,180 @@
+RESOURCE_NAME = 'resource'
+
+
+class TypeKey(typing.NamedTuple):
+  """Information about a resource type."""
+  name: str
+  distribution_key: typing.Optional['DistributionKey']
+
+  def type_id(self) -> str:
+    """Returns a unique string identifier for the type."""
+    return json.dumps(self.as_dict(), sort_keys=True)
+
+  def as_dict(self) -> typing.Dict[str, Json]:
+    """Returns a dictionary representation of the distribution key."""
+    dist_info = (
+      self.distribution_key.as_dict() if self.distribution_key else None)
+    return {
+      'name': self.name,
+      'distribution_key': dist_info}
+
+  @staticmethod
+  def from_dict(d: typing.Dict[str, Json]) -> 'TypeKey':
+    """Instantiate TypeKey from a dictionary."""
+    r: typing.Dict = dict(d)
+    r['distribution_key'] = DistributionKey.from_dict(r['distribution_key'])
+    return TypeKey(**r)
+
+
+class DistributionKey(typing.NamedTuple):
+  """Information about a package where a resource is defined.
+
+  This information (together with qualified class names) is used to identify
+  compatible resources across different versions of a package.
+  """
+  name: str
+  version: str
+
+  def as_dict(self) -> typing.Dict[str, Json]:
+    """Returns a dictionary representation of the distribution key."""
+    return {'name': self.name, 'version': self.version}
+
+  @staticmethod
+  def from_dict(d: typing.Dict[str, Json]) -> 'DistributionKey':
+    """Instantiate DistributionKey from a dictionary."""
+    return DistributionKey(**typing.cast(typing.Dict[str, str], d))
+
+
+@dataclasses.dataclass  # Type descriptors are dataclasses for value-semantics.
+class TypeDescriptor(abc.ABC):
+  """Interface for type descriptors."""
+
+  @abc.abstractmethod
+  def to_json(self) -> Json:
+    """JSON representation."""
+
+  @staticmethod
+  def from_json(json_value: Json) -> 'TypeDescriptor':
+    """Constructs a descriptor from JSON."""
+    if json_value == INT_NAME:
+      return Int()
+    if json_value == STR_NAME:
+      return Str()
+    if json_value == FLOAT_NAME:
+      return Float()
+    if json_value == BOOL_NAME:
+      return Bool()
+    if json_value == BYTES_NAME:
+      return Bytes()
+    if json_value == LIST_NAME:
+      return List()
+    if json_value == DICT_NAME:
+      return Dict()
+    if not isinstance(json_value, dict) or len(json_value) != 1:
+      raise ValueError(
+        f'Expected a dictionary of length 1 for complex type descriptor: '
+        f'{json_value}')
+    if RESOURCE_NAME in json_value:
+      value = json_value[RESOURCE_NAME]
+      if not isinstance(value, dict):
+        raise ValueError(f'Expected a dictionary for resource type descriptor: '
+                         f'{json_value}')
+      return ResourceType(TypeKey.from_dict(value))
+    raise ValueError(f'Unknown type descriptor: {json_value}')
+
+@dataclasses.dataclass
+class Int(TypeDescriptor):
+  """Describes an int value."""
+
+  def to_json(self):
+    return INT_NAME
```

Comment:
Recommend moving these names into the type dataclasses which would make the logic a bit more self contained. So you could do:

@dataclasses.dataclass
class Int(TypeDescriptor):
  """Describes an int value."""
  NAME: ClassVar[str] = 'int

  def to_json(self):
    return Int.NAME

Reply:
Done